[Android] Improve notification UI.

AkashDhiman opened the pull request:
The first 3 commits are updating some related code.

The next 4 commits are adding/removing some data structures to facilitate building up of this UI.

The last 2 commits are involved in changing the notification structure
- first one makes it so that the app shows 1 notifications per conversation.
- second one groups the notification.

Summary notification is not yet implemented correctly in this PR, and will show incorrect information (gives summary for all notifications on a summary notification for 1 realm) when API < 24 and user is logged in with multiple realms.

Some methods may no longer be needed, yet are not removed because they may prove useful in implementing other features.

This skips #4633 and fixes part of #2691, part of because summary notification is not yet implemented properly.

some screenshots of the same:
## Summary notification:
![WhatsApp Image 2021-07-01 at 7 55 32 PM (4)](https://user-images.githubusercontent.com/40268170/124141037-aa35a200-daa6-11eb-8447-c520a27a8e0b.jpeg)

## Summary expanded
![WhatsApp Image 2021-07-01 at 7 55 32 PM (3)](https://user-images.githubusercontent.com/40268170/124141028-a86bde80-daa6-11eb-9f65-4ddc57dfbd0b.jpeg)

## Individually Expanded
![WhatsApp Image 2021-07-01 at 7 55 32 PM (2)](https://user-images.githubusercontent.com/40268170/124141021-a6a21b00-daa6-11eb-90b0-9009857c8613.jpeg)

## Message from Another realm
![WhatsApp Image 2021-07-01 at 7 55 32 PM](https://user-images.githubusercontent.com/40268170/124141015-a4d85780-daa6-11eb-980c-cb37b94f3c29.jpeg)

## Future Work
- Provide a proper summary notification.
- Add action buttons to mark as read, mute topic and reply from notification.
- Add custom sound.
- Remove redundant code.

gnprice commented:
https://github.com/zulip/zulip-mobile/pull/4842

Thanks @AkashDhiman! This looks like a lot of good changes. Looking forward to having this implemented.

    Caveats: There is currently no good way to construct title for
    groupPm type conversation. Ideally it should be the names of all the
    people in the conversation, but we only have their UserId in the
    message payload.

Yeah. This is an example of where to really do a great job, we need some of the data that the app has in Redux and that we don't currently have a great way to share for the notifications code to use.  This is related to #4841, in that it's another motivation for wanting to change our model for storing data.

I wouldn't want to let that block all the other progress this PR is about, though. For this PR, maybe just have it say something like "Group PM"? (And maybe plural if there's more of them? Though getting that right in a translatable way is [a bit complicated](https://developer.android.com/guide/topics/resources/string-resource#Plurals).) From your first screenshot, it looks like if we do that, then the name of the first sender will come right after that, so when there are several PM threads they'll still be possible to tell apart.

---

I like the switch to `NotificationCompat`! In general the "compat" libraries, aka AndroidX, are good to use whenever applicable, for the reason you say.

The fact that it simplifies `getNotificationManager` to where we can just inline it everywhere is a nice bonus. And I appreciate that you separate that inlining into its own commit :slightly_smiling_face:

---

The new `Conversation` data class makes a lot of sense.

> notificationId […] (usually set to the first messageId of the conversation.)

It looks like this becomes the notification ID we use with the Android SDK. What constraints does that ask for on how we pick and use the notification ID?

I think grouping by `realmId` won't quite work, though, for a couple of reasons:
* For most servers that aren't Zulip Cloud, it'll all be the same value -- the ID you get for the first realm you set up on the server. (Probably 2? I think an internal realm uses realm ID 1.) So if you get notifications from several different self-hosted servers, they'll get grouped together.
* If you have several accounts in the same org/realm, the notifications you get for those will be grouped together. This isn't a common use case outside of development -- but it often is very helpful for development, so it's worth spending some effort to make several accounts on the same org work similarly to several accounts in different orgs.

Those two reasons are what the `Identity` type is there to handle. Using `realmUri` instead of `realmId` takes care of the first one. Using `userId` takes care of the other.

Probably a step that would be helpful for using `Identity` here is to give it a method that returns a unique string key based on its contents. Mostly it would just take `realmUri` and `userId.toString()`, and concatenate them separated by some character that can't appear in a URL.

---

Relatedly, if we store the conversations in `ConversationMap` keyed only on things like the stream name, topic, and user IDs, then we have a bug where they can collide if those happen to match. For example, if a user is present on several self-hosted servers, the users they're talking to there are likely to have small user IDs that may well collide. And a lot of stream names like `#general` are very common, and people fairly often use short topics that are pretty generic and can easily collide.

That's a bug we have today, so it's not exactly a regression -- but it becomes more conspicuous with these improvements to the UI that make it easier to get more information from it :stuck_out_tongue: . So that's why the sketch in `ByConversationMap` is there.

---

In the first commit:

    notif [nfc]: Update deprecated code concerning notification token.
    
    The deprecated code was not causing any noticable bugs, but android
    studio was prompting that we should fix it, and its good to have up
    to date code regardless.
    
    One benefit of this specific commit is that it removes some code.

This sounds great! I like deleting code :smile: There's a couple of things you can do to make this change easier for me to review and merge:

* Is there documentation somewhere you can link to about this API change? I'd hope that the Android Studio notice about it would link to something like that.

* There's also a couple of changes in `NotificationsModule.java` related to `FirebaseInstanceId.getInstance().getToken()`, and at least one of them looks separate from the change to drop `InstanceIDListenerService`. Can that (or those) be a separate commit from the `InstanceIDListenerService` change? That way it can get a separate commit message with its own explanation.

AkashDhiman commented:
@gnprice Ready for a re-review 🙂, I recommend merging #4938 as this PR has it as its base.
I have implemented this with a new logic where we don't have to rely on `conversationMap` to store past notifications so its quite different from the past revision.
There is a regression involving handling of `RemoveFcmMessage` which I am looking into. I have added TODOs wherever I think more work is needed.

AkashDhiman commented:
pushed again changing 1 small thing in  (5e8c31b: `notif: Update Notification UI in android.`),
call to notify on non summary notification had value of TAG set to `groupKey,` now changed to `conversationKey`.
Fixed a typo "naem" -> "name" in a comment.

gnprice commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -65,129 +71,156 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
+        }
+    }
+    val groupKey = "${fcmMessage.identity?.realmUri.toString()};${fcmMessage.identity?.userId}"
+    val conversationKey = "$groupKey;$title"
```
And let's pull this out as its own helper function too, this one taking the `FcmMessage`. This helper's job is:
* The result should be constant for a given conversation.
* It should be distinct for distinct conversations.

That points at a few improvements we can make, relative to the solution of using the user-facing title as it's defined above:
* For type `Recipient.Pm`, use `sender.id` instead of `sender.fullName`. That way (a) it doesn't get messed up if they change their name, (b) there's no possibility of the key getting confused because the delimiter character appeared in the data.
  * Before relying on `sender.id`, we should make it non-null, the same way as `identity` as discussed above; our notes say this also dates to server v1.8.0.
  * If we didn't reliably have `sender.id`, then `sender.email` would be a next-best solution.
* For `Recipient.GroupPm`, `recipient.pmUsers` is perfect. Best to skip the `getString`, in fact, so this internal data structure's workings aren't complicated by interactions with a translated user-facing string.
* For `Recipient.Stream`... ideally we'd use the stream ID instead of the stream name, but it looks like we don't currently have that in the data. For now, I guess we can settle for a TODO comment in the code saying to switch to stream ID. And for a delimiter, let's use the character `\x00`, for the same reason as we do that in `keyFromNarrow` in `narrow.js`:
  ```js
      // '\x00' is the one character not allowed in Zulip stream names.
      // (See `check_stream_name` in zulip.git:zerver/lib/streams.py.)
      topic: (streamName, topic) => `topic:s:${streamName}\x00${topic}`,
  ```
* Oh also, looking at `keyFromNarrow` reminds me: for these conversation keys let's also have a different fixed prefix for each of the three types of recipient. That way, each of the three different ways we format a conversation key always has a distinct prefix, so it's easy to be sure that they'll never accidentally collide with each other.

AkashDhiman replied:
done, using `\u0000` instead of `\x00`, kotlin was not letting me use `\x`, I hope that's fine

gnprice replied:
Sure -- that should be another way (and I guess the only way, in Kotlin) of spelling the same thing.

gnprice reviewed (commented):
Thanks for the revisions! Inline comments below.

I really like the prospect of having the system effectively track the different messages we're showing notifications for, rather than having to do that ourselves. That is a nice simplification.

I'm guessing that's also the source of the challenge with maintaining the functionality of `RemoveFcmMessage`, i.e. to remove a notification from the UI when you go and read the relevant message from some other client. In the MessagingStyle doc:
https://developer.android.com/reference/androidx/core/app/NotificationCompat.MessagingStyle
I see there's an `addMessage`, but nothing with a name like `removeMessage`.

Here's a couple of ideas -- but I'm curious also what you've looked at so far, and maybe a thread on #mobile-team is the best place for detailed discussion:
* There is `getMessages`, which returns a `List`. The `List` interface has `List#remove`, but it's optional -- some `List` implementations will throw on it. So possibly we could use that, though it'd be a bit shady because the API of `MessagingStyle` doesn't really guarantee it'll work (even if empirically it currently does.)
* It seems like the sticky case here is when the user has read some but not all of the messages that are in a given UI notification (so, that are all in a given conversation.) Once they read *all* of the messages in the conversation, we can just clear that whole notification.
  * I think it's actually a pretty fine UX if nothing in the notification UI changes until you read all of the relevant messages for a given conversation. So if we accept that, then that leaves the problem of how to know when that's happened.
  * I think a pretty acceptable approximation would be: once you read the *latest* of the messages in a given conversation, we presume you've read all of them, and clear that conversation's UI-notification. That may help keep this implementation nice and simple.
  * But I'd also be open to solutions that detect exactly when you've read all of them, instead of the one latest message, if you see a nice way to arrange that.

gnprice commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -65,129 +71,156 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
-    builder.addAction(action)
+    val summaryNotification = createSummaryNotification(context, fcmMessage, groupKey)
 
-    val soundUri = getNotificationSoundUri(context)
-    val audioAttr = AudioAttributes.Builder()
-        .setUsage(AudioAttributes.USAGE_NOTIFICATION).build()
-    builder.setSound(soundUri, audioAttr)
```
This is one chunk that I notice doesn't have a counterpart in the new code.

Looking at the API docs, I see that it has no effect on Android 8 O and up!
https://developer.android.com/reference/androidx/core/app/NotificationCompat.Builder#setSound(android.net.Uri)
So it wasn't doing us a lot of good. But: that's because we're supposed to be calling the corresponding method on the notification *channel*'s builder instead:
https://developer.android.com/reference/androidx/core/app/NotificationChannelCompat.Builder#setSound(android.net.Uri,%20android.media.AudioAttributes)

So it seems like that's something we neglected to take care of when we started using a notification channel a few years ago. But seeing this code and deleting it gives us a nice reminder of how we should be doing it. Would you add a few lines to our `createNotificationChannel` to fix that?

The very cleanest thing would be to make sure we continue to do it on the notification itself, for the sake of pre-Android 8 devices, as well as on the channel. But at this point those devices are well into the single-digit percent (they [were 9.0%](https://chat.zulip.org/#narrow/stream/48-mobile/topic/platform.20versions/near/1012192) almost a year ago), so graceful degradation is okay, and I'd be happy with a version that only does it on the channel.

AkashDhiman replied:
I have added `audioAttr` use in channel as defined previously for notification builder.
`NotificationCompat.Builder` doesn't allow me to call `setSound` with `audioAttr` (there is no function with that signature in `NotificationCompat.Builder`) so i don't think i can easily add it there.

gnprice reviewed (commented):
(One more comment on a bit I just noticed.)

gnprice commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -2,44 +2,50 @@
+        // value, hence using "SupressLint".
+        val channel =
+            NotificationChannel(CHANNEL_ID, name, NotificationManagerCompat.IMPORTANCE_HIGH).apply {
+                description = context.getString(R.string.notification_channel_description)
```
Hmm -- can you post a screenshot of how this description appears in the UI? That would help me wordsmith the text for the description.

I'm not actually seeing anything that looks like this when I browse around the notification settings for other apps on my phone, which is making me suspect that my phone (Android 11 on a Pixel) just ignores the description. I also don't see anything that looks like it in the handy guide doc you linked:
https://developer.android.com/training/notify-user/channels

AkashDhiman replied:
<img width="331" alt="Screenshot 2021-08-19 at 10 12 20 PM" src="https://user-images.githubusercontent.com/40268170/130109223-5abdf968-ed5b-4ac9-af2a-f44806983041.png">

At the very end it is the grayed out message. This is accessible on tapping the notification channel name from the app setting.

gnprice replied:
Fascinating, thanks.

Do you see this on any other apps you happen to have lying around? I just looked at this part of the UI for several well-known apps -- including Facebook Messenger, which has an extensive set of notification categories, so they've clearly paid attention to this area of the API, and also including Android Messages, which ditto plus it's actually from the Android group at Google -- and I don't see a grayed-out message like that at all. There's the "Advanced" section, and then nothing after that. (Also, just to confirm: what Android version are you on?)

If other app makers, including the Google Android group itself, aren't using this feature, then we clearly don't need to feel obligated to do so, and I'm inclined to skip it. I think the category's name itself can convey all the same information -- perhaps we should rename it "All messages" to be clearer -- and the way the UI ends up presenting this description text feels pretty out-of-context and not likely to be helpful.

gnprice replied:
I looked again at the settings UI where this name appears, and I think "Messages" works well as the name and "All messages" would be unnecessarily redundant.

A couple of quick comparison points:
* FB Messenger has "Chats and calls".
* Android Messages has "Incoming messages".

gnprice commented:
OK, I've merged 6/13 commits from the initial part of this PR!

c2087fa33 android notif [nfc]: Remove `InstanceIDListenerService.java`.
78a2b46bd android notif [nfc]: Update deprecated code in `getToken`.
5763242cc android notif [nfc]: Use compat version for notification modules.
507475d4a android notif [nfc]: Remove `getNotificationManager`.
6320ae324 android notif [nfc]: Remove `ByConversationMap`.
f666c414e android notif: Better name channel; enable lights and vibration.

I fixed up the conflict-resolution error mentioned at the top of my review today. I also dropped the channel description for now because I'm not sure how that will show up; see comment just above this.

The only other change is just that I tweaked the summary lines to say "android notif". That slightly more specific context is helpful for interpreting summaries like "Use compat version for notification modules." -- aha, it's Android, so that's the Android compat library aka AndroidX -- and it's also helpful just for imagining what kind of code this is touching, because the Android-native code is fairly distinct in some ways from the JS code.

Looking forward to the next revision, and to getting to merge the main changes here soon :smile:

AkashDhiman commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -65,129 +71,156 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
+ */
+private fun getActiveNotification(context: Context, conversationKey: String):
+    Pair<Int?, Notification?> {
+    // activeNotifications are not available in NotificationCompatManager
+    // Hence we have to use instance of NotificationManager.
+    val notificationManager =
+        context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager?
```
> Do you know what the story is there?

Not sure about this, I couldn't find any reference for why this is the case.

On a side note it is also possible to get active notification via [NotificationListenerService#getActiveNotifications](https://developer.android.com/reference/android/service/notification/NotificationListenerService#getActiveNotifications(java.lang.String[])) so they could have nicely consolidated this for most of the android versions. This is how one can get `activeNotifications` in API < 23.

AkashDhiman commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -65,129 +71,156 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
+    if (activeStatusBarNotifications != null) {
+        for (statusBarNotification in activeStatusBarNotifications) {
+            val notification = statusBarNotification.notification
+            if (notification.extras.getString("conversationKey") == conversationKey) {
```
doing it this way in the new revision thanks for that!

AkashDhiman commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -65,129 +71,156 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
+    fcmMessage: MessageFcmMessage,
+    groupKey: String
+): NotificationCompat.Builder {
+    val realmUri = fcmMessage.identity!!.realmUri.toString()
```
changed realmUri and Identity and Sender.id to be Non Nullable in the class.

AkashDhiman commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -65,129 +71,156 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
+
+    val builder = NotificationCompat.Builder(context, CHANNEL_ID)
+    builder.apply {
+        color = Color.rgb(100, 146, 254)
```
done, Shouldn't this ideally be in `colors.xml`, so that we only have to update it in one place.

gnprice replied:
Hmm, yeah, good idea! That'd be a great change to make either before or after the commit where you add a new place we use the color.

AkashDhiman commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -65,129 +71,156 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
-        //Only one 1 notification therefore no using of big view styles
-        if (totalMessagesCount > 1) {
-            builder.setContentTitle("${fcmMessage.sender.fullName} ($totalMessagesCount)")
+private fun createSummaryNotification(
```
> How did you find that it's required by Android even on newer versions?

I tested a build without summary Notification, specifically without the notify call to summaryNotification, observation is that notification do occur but they don't group together, even in android 7+. It seems that this notify call triggers the actual grouping.

Also while they don't appear completely in Android 7+, (since they get overridden by the summary notification built by the Android OS) they still have some visual influence on how the summary notification looks, for instance, setting
`summaryText` shows as the [header text](https://material.io/design/platform-guidance/android-notifications.html#anatomy-of-a-notification) in summary notification.

gnprice replied:
> I tested a build without summary Notification, specifically without the notify call to summaryNotification, observation is that notification do occur but they don't group together, even in android 7+. It seems that this notify call triggers the actual grouping.

Hmm fascinating -- thanks for determining that. Seems like a bug [in the docs](https://developer.android.com/training/notify-user/group) that they say it's optional, then.

I guess in the example code says:
```kt
val summaryNotification = NotificationCompat.Builder(this@MainActivity, CHANNEL_ID)
        .setContentTitle(emailObject.getSummary())
        //set content text to support devices running API level < 24
        .setContentText("Two new messages")
        // […]
```
which makes it sound more like the actual situation you discovered, that the summary notification itself is required but the text is only for older devices.

We did [find when I looked recently](https://chat.zulip.org/#narrow/stream/48-mobile/topic/platform.20versions/near/1254509) that indeed Android 6 and older is down to a small enough sliver of our userbase that it makes sense to drop support for it as soon as there's something where supporting it would make things more complicated for us. So I think we won't go and add a summary text, and instead will just drop support and increase the minimum to Android 7.

AkashDhiman commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -65,129 +71,156 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
-    builder.setDefaults(Notification.DEFAULT_VIBRATE or Notification.DEFAULT_LIGHTS)
+private fun updateNotification(
+    context: Context, fcmMessage: MessageFcmMessage) {
+    val user = Person.Builder().setName("You").build()
```
> It is, right?

Yes its the person using the app through this client.

gnprice replied:
> * "You" should be a string resource so it can get translated.

bump

AkashDhiman commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -65,129 +71,156 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
-    // return Uri.parse("${ContentResolver.SCHEME_ANDROID_RESOURCE}://${context.packageName}/${R.raw.zulip}")
-    return Settings.System.DEFAULT_NOTIFICATION_URI
+        updateNotification(context, fcmMessage)
+    } // TODO handle case for RemoveFcmMessage
```
This is the only thing that remains to be implemented properly, I have made a "hacky" version but it has few problems, see 1b9d2e7 for detail.

AkashDhiman commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -65,129 +71,156 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
+            ?: NotificationCompat.MessagingStyle(user)
+    }
+    messagingStyle
+        .setConversationTitle(title)
+        .setGroupConversation(isGroupConversation)
```
I am assuming I don't have to do anything for this then right? (its unchanged in the new revision)

gnprice replied:
Yep, that's right!

gnprice commented:
Thanks for the revision!

Two high-level comments on the "remove" case:

```kt
    // TODO figure out a way to cancel individual group notifications
    // misbehaves with multiple realm, shows empty notification.
    // This cancels all group Notification when no other notification
    // is present.
    if (Build.VERSION.SDK_INT >= 24) {
        if (getActiveNotification(context)?.all { statusBarNotification -> statusBarNotification.isGroup } == true) {
            NotificationManagerCompat.from(context).cancelAll()
        }
    }
```

So this is scanning through all the existing notifications and checking whether it's the case that they're all… well, if this works then it's checking that they're all group summaries, I guess. (The [docs actually say](https://developer.android.com/reference/android/service/notification/StatusBarNotification#isGroup()) that `isGroup` is true "if this notification is part of a group", but it doesn't make totally clear what that's supposed to mean.) If they are, then we must have removed the last of the underlying individual notifications, and we can clear out the groups.

To get the desired behavior, it seems like we want to do something pretty similar, except instead of checking whether there are any individual (i.e., non-group-summary) notifications at all, we want to check whether there are any of those for each group key. So for example we could do a scan through these and build up a `Set` of group keys to keep, and then another scan to find group-summary notifications that are for empty groups and should be cancelled.

```kt
    for (fcmMessageId in fcmMessage.messageIds) {
        val activeNotifications = getActiveNotification(context) ?: return
        for (statusBarNotification in activeNotifications) {
            val notification = statusBarNotification.notification
            val messageId = notification.extras.getInt("zulipMessageId")
            if (fcmMessageId == messageId) {
```

We can make this linear instead of quadratic by turning `messageIds` into a `Set`, and then making just one linear scan where we check membership in that. … Oh hey, actually we already made a `Set` at parse time :smile: -- `fcmMessage.messageIds` is a `Set<Int>`. So we can write it like this:
```kt
        val activeNotifications = getActiveNotification(context) ?: return
        for (statusBarNotification in activeNotifications) {
            val notification = statusBarNotification.notification
            val messageId = notification.extras.getInt("zulipMessageId")
            if (fcmMessage.messageIds.contains(messageId)) {
```

AkashDhiman commented:
@gnprice did the required changes, I think the notifications are getting dismissed as expected now, you can review it again.
I have also added the string I thought was most suited for group pm in this revision, please see if that's appropriate.

gnprice commented on android/app/.project:
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
It looks like this file was included unintentionally.

Want to add a line for it to our gitignore? Probably it should go in this block:
```
# VS Code: default Java extension
/android/.project
/android/.settings/
```
which already has a similar line. Though if that comment doesn't describe how you ended up with this file, then we'd update the comment too.

gnprice commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -69,124 +76,215 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
-        NotificationManagerCompat.from(context).cancelAll()
-        return
+fun removeNotification(context: Context, fcmMessage: RemoveFcmMessage) {
+    val statusBarNotifications = getActiveNotification(context) ?: return
+    val groupKey = extractGroupKey(fcmMessage.identity)
+    for (statusBarNotification in statusBarNotifications) {
+        val notification = statusBarNotification.notification
+        val messageId = notification.extras.getInt("zulipMessageId")
+        if (fcmMessage.messageIds.contains(messageId)) {
+            NotificationManagerCompat.from(context).cancel(statusBarNotification.tag, statusBarNotification.id)
```
Oh also, in that first comment a link to https://github.com/zulip/zulip-mobile/pull/4842#pullrequestreview-725817909 would be helpful, to give a pointer back into the history of our reasoning.

AkashDhiman replied:
changed.

gnprice commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -69,124 +76,215 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
+    val summaryNotification = createSummaryNotification(context, fcmMessage, groupKey)
+
+    NotificationManagerCompat.from(context).apply {
+        // here notifications are identified by TAG + notificationId
+        // simply using notificationId is not sufficient and will
+        // fail in case of different account in same realm.
+        notify(groupKey, fcmMessage.identity.realmId, summaryNotification.build())
+        notify(conversationKey, notificationId, builder.build())
```
Looking again at these notification "id"s, the second argument to each of these `notify` calls:

Do we ever use the fact that the summaries have the realm ID as their notification ID? I think we don't; and I think if we did, that would probably be a bug, because realm IDs are per-server and in fact are highly likely to collide between any two servers other than Zulip Cloud (because they'll probably both be the only realm on the server.) If we don't, how about just using a constant ID, like 1? That would help make it clear that we're relying entirely on the tags.

Similarly, do we ever use the fact that the notification IDs for individual conversations vary, or that they're Zulip message IDs? I think we don't, in which case it'd similarly be good to just use a constant. It can even be the same constant -- we've designed these keys so they don't collide between the two types, either.

AkashDhiman replied:
using NOTIFICATION_ID, like the main branch here.

gnprice reviewed (commented):
Thanks @AkashDhiman for the revision! This is looking great. Comments below, and above because GitHub does that when I reply on a previous thread, and a few other comments further above in replies on other threads:
https://github.com/zulip/zulip-mobile/pull/4842#discussion_r719123898
https://github.com/zulip/zulip-mobile/pull/4842#discussion_r719129793
https://github.com/zulip/zulip-mobile/pull/4842#discussion_r719130444
https://github.com/zulip/zulip-mobile/pull/4842#discussion_r719137940

I merged #4967 a little while ago, and I think that should mean that when rebasing you'll get to drop this commit:
28dadc9cf notif: Make some MessageFcmMessage properties Non Nullable.
because that took care of ensuring we always have those properties.

AkashDhiman commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -69,124 +76,215 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
+        val messageId = notification.extras.getInt("zulipMessageId")
+        if (fcmMessage.messageIds.contains(messageId)) {
+            NotificationManagerCompat.from(context).cancel(statusBarNotification.tag, statusBarNotification.id)
+            break
```
yes, there shouldn't be a need for break here.

AkashDhiman commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -69,124 +76,215 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
+        // statusBarNotification.groupKey also contains package name
+        // in the beginning and hence doesn't match the groupKey we
+        // set in notification builder; therefore we use endsWith.
+        if (statusBarNotification.groupKey.endsWith(groupKey)) {
```
using `statusBarNotification.notification.group == groupKey`, the `groupKey` in `statusBarNotification` contains "ranking" which may also have a number concatenated at the beginning which I am not sure how to compute.

gnprice replied:
Huh interesting. Well, glad that solution works.

AkashDhiman commented:
2 small things that remain from last time's review:
- [ ] write the `brandColor` in `colors.xml` https://github.com/zulip/zulip-mobile/pull/4842#discussion_r719123898
- [ ] rename the channel from "Messages" to "All Messages" https://github.com/zulip/zulip-mobile/pull/4842#discussion_r719137940

Between this revision and the previous one, The layout seems to have changed a bit, but I am not sure why, hence before merging I recommend we figure out the reason and try to move back to that layout, It has larger text for heading and a default group conversation icon.
**Previous Layout**:
<img width="583" alt="Screenshot 2021-11-04 at 9 59 50 PM" src="https://user-images.githubusercontent.com/40268170/140379934-b8c7d55c-21ba-42f3-8319-39fa6426539a.png">

**New Layout (Expanded Group Conversation)**
<img width="272" alt="Screenshot 2021-11-04 at 10 16 09 PM" src="https://user-images.githubusercontent.com/40268170/140382585-be670c07-74fe-4647-91a1-48cd7df4d464.png">
**New Layout (Closed Group Conversation)**
<img width="284" alt="Screenshot 2021-11-04 at 10 15 59 PM" src="https://user-images.githubusercontent.com/40268170/140382570-49fb7355-a7d2-41ca-a4c2-9073e49da96d.png">

gnprice commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -33,6 +37,8 @@ val ACTION_CLEAR = "ACTION_CLEAR"
 val EXTRA_NOTIFICATION_DATA = "data"
 
 fun createNotificationChannel(context: Context) {
+    val audioAttr: AudioAttributes = AudioAttributes.Builder()
+    .setUsage(AudioAttributes.USAGE_NOTIFICATION).build()
```
nit: second line should be indented

(really by whatever the autoformatter wants to do, probably, but if doing it without an autoformatter I'd go for 4 spaces)

gnprice commented on android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt:
```diff
@@ -70,124 +77,211 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
+
+    val summaryNotification = createSummaryNotification(context, fcmMessage, groupKey)
+
+    NotificationManagerCompat.from(context).apply {
+        notify(groupKey, NOTIFICATION_ID, summaryNotification.build())
+        notify(conversationKey, NOTIFICATION_ID, builder.build())
```
This could use a line or two of comment saying that we rely entirely on the tags to distinguish notifications, and so the "IDs" can be a constant. I.e. summarizing the reasoning at https://github.com/zulip/zulip-mobile/pull/4842#discussion_r719157582 .

gnprice reviewed (commented):
Thanks for the revision! The code looks good; just two small comments below. I should also play with using the new version.

In addition to the two small open items you mention, I see one other:
- [ ] We should take this as a prompt to drop Android 6 support (i.e. bump the minimum to Android 7): https://github.com/zulip/zulip-mobile/pull/4842#discussion_r719129793

---

> Between this revision and the previous one, The layout seems to have changed a bit, but I am not sure why, hence before merging I recommend we figure out the reason and try to move back to that layout, It has larger text for heading and a default group conversation icon.

Hmm, yeah, puzzling. I'm not sure if I like the old or the new version better -- it looks like the new one makes a bit more space for the message content (when closed/unexpanded), which is good. But whichever way we end up deciding we prefer, I definitely agree it'd be good to understand what caused the difference, so that we can be confident we understand the APIs we're using.

I think you said on a call that you can still reproduce the old behavior if you build from the previous revision of your branch, on the same device where you get the new behavior with this new revision. Is that right?

What's the commit ID of the previous revision where you get that previous behavior? (Just so I can try to reproduce and know I'm trying the same thing you are.) The previous version that I have fetched is 081669a78, and before that 1b9d2e714.

Probably the next step I'd try for narrowing it down is to take the old revision, and just rebase it onto the same commit from main that the new revision is based on, and see if that has the old or the new behavior.

Then if that has the new behavior, bisect what version in main to rebase onto. If it has the old behavior, then try to bisect the differences between the (rebased) old revision and the new revision.

Hmm, one possible explanation occurs to me: the change in targetSdkVersion. Previously we targeted Android 10, which meant that Android 11+ would treat us in some respects as having been designed for Android 10 and accommodate what we would have expected from testing on Android 10. Now we'll be treated as having been designed for Android 11. So that hypothesis can be tested by rebasing first onto the commit before we made that change, and confirming it's still the old behavior, then rebasing just one commit further and confirming it's now the new behavior.

If that is it, then there may or may not be a knob we can turn to get the old behavior, but it'll certainly help us find the relevant bit of docs.

AkashDhiman commented:
> What's the commit ID of the previous revision where you get that previous behavior?

I see the previous layout in https://github.com/zulip/zulip-mobile/commit/081669a78898c3a005eff19d9cb593d1be3621bd

AkashDhiman commented:
I have figured out the reason for the layout change, it's due to the change in `targetSdkVersion` from 29 to 30 30b4d6a.

I made a fresh app in android to test this as well and it checks out.

It seems that this cannot be changed by us then?

gnprice commented:
OK, good to know. That confirms my guess above, then.

I took a look at the list of behavior changes for apps targeting Android 11 (i.e. `targetSdkVersion` set to 30 instead of lower):
https://developer.android.com/about/versions/11/behavior-changes-11
and don't see anything about this there.  So it seems the change isn't well documented.

I think the new layout actually works pretty well for us by giving more room for the stream name and topic. So I'm inclined to let that be.

---

I just did some playing with this version, sending myself messages from a couple of test accounts in different kinds of conversations. It worked very nicely! I am looking forward to getting this merged.

There's one thing I found that I'd like to fix before merge:
- [ ] When all the messages are read in a given conversation, the conversation's notification disappears nicely. But when the last conversation's notification disappears (for a given account), there's still an awkward sort of header notification there -- just the name of the app and the URL of the org.

Then a few things that would be good as followups, and may require some server-side changes and/or infrastructural work on our side. These are probably best converted into issues.
* The heading for the whole group identifies the account by the org's URL. Probably this should be the org name instead?
  * For I guess one comparison: the notifications I get from various Google services have my Google account email there.
  * Would be good to look at what Discord and Slack do here.
* For group PMs, the conversation heading is like "from Greg Price with 1 other". This is a known issue we discussed above; the difficulty is that we have the other recipients' user IDs but not display names.

AkashDhiman commented:
@gnprice done, I checked the new revision and it should work fine now; also added `brandColor` in `colors.xml`.

AkashDhiman commented:
There was an unused library import, pushed again.

gnprice commented:
Thanks! Glad that was a straightforward fix. (And [here in chat](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Notification.20UI/near/1278077) is where you debugged it, for cross-reference.)

This looks good; I'll merge.

The remaining open items were:
* We'd discussed changing the name of the notification channel, from "Messages" to "All messages": https://github.com/zulip/zulip-mobile/pull/4842#discussion_r719137940 I looked back at that just now, and I think it's better as it is.
* This change gave us a prompt to drop Android 6 support, bumping the minimum to Android 7: https://github.com/zulip/zulip-mobile/pull/4842#discussion_r719129793 We should probably do that, but it sounds like the symptom on Android 6 isn't terrible, so we don't need to do so before merge.
* There's a couple of followup items here: https://github.com/zulip/zulip-mobile/pull/4842#issuecomment-963481595

Thanks again @AkashDhiman for all your work on this!
